File: haskell/grains/src/Grains.hs
module Grains
  ( square
  , total
  ) where

import qualified Data.List as List

squares :: [Integer]
squares = 1 : iterate (* 2) 1

square :: Integer -> Maybe Integer
square n
  | n > 64 = Nothing
  | n > 0 = Just (List.genericIndex squares n)
  | otherwise = Nothing

total :: Integer
total = (sum . take 65) squares - 1

File: exercism/elixir/grains/grains.exs
defmodule Grains do
  @doc """
  Calculate two to the power of the input minus one.
  """
  @spec square(pos_integer) :: pos_integer
  def square(number) do
    hd(squares number)
  end

  defp squares(number) do
    Stream.iterate(1, &(&1 * 2))
    |> Stream.take(number)
    |> Enum.drop(number - 1)
  end

  @doc """
  Adds square of each number from 1 to 64.
  """
  @spec total :: pos_integer
  def total do
    Enum.sum(squares 65) - 1
  end
end
